src/hotspot/share/gc/serial/defNewGeneration.cpp

```diff
@@ -621,7 +625,12 @@ void DefNewGeneration::ref_processor_init() {
   assert(_ref_processor == nullptr, "a reference processor already exists");
   assert(!_reserved.is_empty(), "empty generation?");
   _span_based_discoverer.set_span(_reserved);
-  _ref_processor = new ReferenceProcessor(&_span_based_discoverer);    // a vanilla reference processor
+  _is_alive_closure.init(this);
+  _ref_processor = new ReferenceProcessor(&_span_based_discoverer,
+                                          1,                        // mt processing degree
+                                          1,                        // mt discovery degree
+                                          false,                    // concurrent_discovery
+                                          &_is_alive_closure);
```

Comment:
This change is not correct. The is_alive_non_header closure for Serial's
reference processor should remain nullptr (as it was before). Setting it here
does nothing but add a useless call in RP::discover_reference that will always
return false (allowing discovery to proceed).

(1) During oop iteration we encounter a young Reference, calls oop_oop_iterate_discovery
(2) oop_oop_iterate_discovery calls try_discover.
(3) try_discover examines the referent.
(3a) referent is nullptr or already marked => no discovery.
(3b) call RP::discover_reference
(4) RP::discover_reference calls the is_subject_to_discovery closure
(4a) obj is old => no discovery
(4b) obj is young => continue discovery check
(5) RP::discover_reference calls is_alive_non_header closure if it exists

The closure being provided will return true if obj is old or marked.  But
we've already filtered out both of those cases above, so the closure will
always return false, so the discovery process should proceed.

Reply:
`is_subject_to_discovery` checks on the reference, while `is_alive_non_header` checks on the referent.

Without this change to ref-processor, the newly added assert in `KeepAliveClosure` (part of this PR) would fail.